codersdk/metrics.go

```diff
@@ -0,0 +1,145 @@
+package codersdk
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/http/cookiejar"
+	"time"
+
+	"golang.org/x/xerrors"
+	"nhooyr.io/websocket"
+	"nhooyr.io/websocket/wsjson"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/agent"
+	"github.com/coder/retry"
+)
+
+type CloseFunc func() error
+
+func (c CloseFunc) Close() error {
+	return c()
+}
+
+// AgentReportStats begins a stat streaming connection with the Coder server.
+// It is resilient to network failures and intermittent coderd issues.
+func (c *Client) AgentReportStats(
+	ctx context.Context,
+	log slog.Logger,
+	stats func() *agent.Stats,
+) (io.Closer, error) {
+	serverURL, err := c.URL.Parse("/api/v2/metrics/report-agent-stats")
+	if err != nil {
+		return nil, xerrors.Errorf("parse url: %w", err)
+	}
+
+	jar, err := cookiejar.New(nil)
+	if err != nil {
+		return nil, xerrors.Errorf("create cookie jar: %w", err)
+	}
+
+	jar.SetCookies(serverURL, []*http.Cookie{{
+		Name:  SessionTokenKey,
+		Value: c.SessionToken,
+	}})
+
+	httpClient := &http.Client{
+		Jar: jar,
+	}
```

Comment:
We should really add `Session-Token` header auth for cli. Once we add CSRF, cookie auth will fail. :cry: